packages/sshnoports/bundles/shell/install.sh

```diff
@@ -0,0 +1,296 @@
+
+# SYSTEM GIVENS #
+define_env() {
+  script_dir="$(dirname -- "$( readlink -f -- "$0"; )")"
+  bin_dir="/usr/local/bin"
+  systemd_dir="/etc/systemd/system"
+
+  user_home="$HOME"
+  user_bin_dir="$user_home/.local/bin"
+  user_sshnpd_dir="$user_home/.sshnpd"
+}
+
+is_root() {
+  [ "$(id -u)" -eq 0 ]
+}
+
+is_darwin() {
+  [ "$(uname)" = 'Darwin' ]
+}
+
+no_mac() {
+  if is_darwin; then
+    echo "Error: this operation is only supported on linux"
+    exit 1
+  fi
+}
+
+root_only() {
+  if ! is_root; then
+    echo "Error: this operation requires root privileges"
+    exit 1
+  fi
+}
+
+# USAGE #
+
+usage() {
+  echo "${arg_zero || 'install.sh' } [command]"
+  echo "Available commands:"
+  echo "at_activate     - install at_activate"
+  echo "sshnp           - install sshnp"
+  echo "sshnpd          - install sshnpd"
+  echo "sshrv           - install sshrv"
+  echo "sshrvd          - install sshrvd"
+  echo "binaries        - install all base binaries"
+  echo ""
+  echo "debug_sshrvd    - install sshrvd with debugging enabled"
+  echo "debug           - install all debug binaries"
+  echo ""
+  echo "all             - install all binaries (base and debug)"
+  echo ""
+  echo "systemd <unit>  - install a systemd unit"
+  echo "                  available units: [sshnpd, sshrvd]"
+  echo ""
+  echo "headless <job>  - install a headless cron job"
+  echo "                  available jobs: [sshnpd, sshrvd]"
+  echo ""
+  echo "tmux <service>  - install a service in a tmux session"
+  echo "                  available services: [sshnpd, sshrvd]"
+}
+
+# INSTALL BINARIES #
+
+install_single_binary() {
+  if is_root; then
+    dest="$bin_dir"
+  else
+    dest="$user_bin_dir"
+  fi
+  mkdir -p "$dest"
+  cp "$script_dir/$1" "$dest/$1"
+  echo "Installed $1 to $dest"
+}
+
+install_base_binaries() {
+  install_single_binary "at_activate"
+  install_single_binary "sshnp"
+  install_single_binary "sshnpd"
+  install_single_binary "sshrv"
+  install_single_binary "sshrvd"
+}
+
+install_debug_binary() {
+  if is_root; then
+    dest="$bin_dir"
+  else
+    dest="$user_bin_dir"
+  fi
+  mkdir -p "$dest"
+  cp "$script_dir/debug/$1" "$bin_dir/debug_$1"
+  echo "Installed debug_$1 to $dest"
+}
+
+install_debug_binaries() {
+  install_debug_binary "sshrvd"
+}
+
+install_all_binaries() {
+  install_base_binaries
+  install_debug_binaries
+}
+
+# SYSTEMD #
+
+post_systemd_message() {
+  echo "Systemd unit installed, make sure to configure the unit by editing $dest"
+  echo "Learn more in $script_dir/systemd/README.md"
+  echo ""
+  echo "To enable the service on next boot:"
+  echo "  sudo systemctl enable $unit_name"
+  echo ""
+  echo "To start the service immediately:"
+  echo "  sudo systemctl start $unit_name"
+}
+
+install_systemd_unit() {
+  unit_name="$1"
+  no_mac
+  dest="$systemd_dir/$unit_name"
+  cp "$script_dir/systemd/$unit_name" "$dest"
+  post_systemd_message
+}
+
+install_systemd_sshnpd() {
+  root_only
+  install_single_binary "sshnpd"
```

Comment:
while systemd units can be installed as a --user unit, it adds a lot of complexity not worth trying to automate.